File: src/AetherNet.Storage/DerivedDataAtRestKeyProvider.cs
// SPDX-License-Identifier: MIT

using System.Security.Cryptography;
using System.Text;

namespace AetherNet.Storage;

/// <summary>
/// <see cref="IDataAtRestKeyProvider"/> that derives a 32-byte AES-256 key
/// from a passphrase and a salt using PBKDF2-HMAC-SHA256. The derived key
/// is cached for the lifetime of the provider so the (relatively expensive)
/// PBKDF2 computation runs exactly once per passphrase/version pair.
///
/// <para>
/// <b>Production iteration count: 600,000.</b> This matches the OWASP 2023
/// recommendation for PBKDF2-HMAC-SHA256 and is the default if no count is
/// supplied. Tests pass a smaller count to keep the suite fast — never lower
/// the default in production code.
/// </para>
///
/// <para>
/// The salt is required, must be at least 16 bytes, and MUST be unique to
/// this device (or this trust boundary). Reusing the same passphrase + salt
/// across devices would let an attacker who recovered the salt from one
/// device decrypt blobs from another — domain-separate by appending an
/// install-id, hardware-id, or randomly generated per-device value.
/// </para>
///
/// <para>
/// To rotate the key, construct a new provider with both the old and new
/// passphrases supplied via <see cref="WithRotation"/> — the old version
/// keeps decrypting historical blobs while new writes use the new key.
/// </para>
/// </summary>
public sealed class DerivedDataAtRestKeyProvider : IDataAtRestKeyProvider
{
    /// <summary>OWASP 2023 recommendation for PBKDF2-HMAC-SHA256.</summary>
    public const int DefaultIterations = 600_000;

    private const int KeyByteLength = 32; // AES-256
    private const int MinimumSaltLength = 16;

    private readonly Dictionary<int, byte[]> _derivedKeys;

    /// <summary>
    /// Construct a single-version provider that derives version 1 from the
    /// supplied <paramref name="passphrase"/> and <paramref name="salt"/>.
    /// </summary>
    /// <param name="passphrase">The user/host passphrase. UTF-8 encoded before derivation.</param>
    /// <param name="salt">At least 16 bytes; should be unique per device.</param>
    /// <param name="iterations">
    /// PBKDF2 iteration count. Defaults to <see cref="DefaultIterations"/>
    /// (the OWASP 2023 recommendation). Tests may pass a smaller value.
    /// </param>
    public DerivedDataAtRestKeyProvider(
        string passphrase,
        byte[] salt,
        int iterations = DefaultIterations)
    {
        ValidateInputs(passphrase, salt, iterations);

        _derivedKeys = new Dictionary<int, byte[]>
        {
            [1] = Derive(passphrase, salt, iterations),
        };
        CurrentVersion = 1;
        Iterations = iterations;
    }

    private DerivedDataAtRestKeyProvider(
        Dictionary<int, byte[]> derivedKeys,
        int currentVersion,
        int iterations)
    {
        _derivedKeys = derivedKeys;
        CurrentVersion = currentVersion;
        Iterations = iterations;
    }

    /// <inheritdoc />
    public int CurrentVersion { get; }

    /// <summary>The PBKDF2 iteration count this provider was constructed with.</summary>
    public int Iterations { get; }

    /// <inheritdoc />
    public byte[]? GetKey(int version)
    {
        return _derivedKeys.TryGetValue(version, out var key) ? key : null;
    }

    /// <summary>
    /// Returns a new provider that adds a freshly derived key under
    /// <paramref name="newVersion"/> (which becomes <see cref="CurrentVersion"/>)
    /// while keeping every existing version available for decryption. Use
    /// during a rotation window: hosts swap the registered provider, run
    /// <see cref="EncryptedKeyValueStore.RewrapAsync"/> across the store
    /// in the background, then drop the old key on the next deploy by
    /// constructing a single-version provider on the new passphrase.
    /// </summary>
    public DerivedDataAtRestKeyProvider WithRotation(
        int newVersion,
        string newPassphrase,
        byte[] newSalt,
        int? iterations = null)
    {
        if (newVersion < 1 || newVersion > 255)
            throw new ArgumentOutOfRangeException(nameof(newVersion),
                "Key version must fit in a single byte (1..255).");
        if (_derivedKeys.ContainsKey(newVersion))
            throw new ArgumentException(
                $"Version {newVersion} already exists in this provider.",
                nameof(newVersion));

        var iters = iterations ?? Iterations;
        ValidateInputs(newPassphrase, newSalt, iters);

        var next = new Dictionary<int, byte[]>(_derivedKeys.Count + 1);
        foreach (var kv in _derivedKeys) next[kv.Key] = kv.Value;
        next[newVersion] = Derive(newPassphrase, newSalt, iters);

        return new DerivedDataAtRestKeyProvider(next, newVersion, iters);
    }

    private static void ValidateInputs(string passphrase, byte[] salt, int iterations)
    {
        ArgumentException.ThrowIfNullOrEmpty(passphrase);
        ArgumentNullException.ThrowIfNull(salt);
        if (salt.Length < MinimumSaltLength)
            throw new ArgumentException(
                $"Salt must be at least {MinimumSaltLength} bytes.",
                nameof(salt));
        if (iterations < 1)
            throw new ArgumentOutOfRangeException(
                nameof(iterations),
                "Iteration count must be positive.");
    }

    private static byte[] Derive(string passphrase, byte[] salt, int iterations)
    {
        // Defensive copy of salt before derivation so caller mutations don't
        // affect the cached key.
        var saltCopy = new byte[salt.Length];
        Buffer.BlockCopy(salt, 0, saltCopy, 0, salt.Length);

        var passphraseBytes = Encoding.UTF8.GetBytes(passphrase);
        try
        {
            return Rfc2898DeriveBytes.Pbkdf2(
                passphraseBytes,
                saltCopy,
                iterations,
                HashAlgorithmName.SHA256,
                KeyByteLength);
        }
        finally
        {
            CryptographicOperations.ZeroMemory(passphraseBytes);
        }
    }
}

File: typescript/src/storage/DerivedDataAtRestKeyProvider.ts
/**
 * {@link DataAtRestKeyProvider} that derives a 32-byte AES-256 key from a
 * passphrase and a salt using {@code scrypt} (Node BCL). The derived key
 * is cached for the lifetime of the provider so the (relatively
 * expensive) derivation runs exactly once per passphrase/version pair.
 *
 * <b>Production cost: N=2^14 (16384), r=8, p=1.</b> This matches the
 * OWASP 2023 guidance for scrypt and is the default if no cost is
 * supplied. Tests pass a smaller cost to keep the suite fast — never
 * lower the default in production code.
 *
 * The salt is required, must be at least 16 bytes, and SHOULD be unique
 * to this device. Reusing the same passphrase + salt across devices
 * would let an attacker who recovered the salt from one device decrypt
 * blobs from another — domain-separate by appending an install-id,
 * hardware-id, or randomly generated per-device value.
 *
 * Rotation: call {@link DerivedDataAtRestKeyProvider.withRotation} to
 * obtain a new provider that adds a freshly-derived key under a new
 * version while keeping every existing version available for
 * decryption.
 *
 * Mirrors the C# {@code DerivedDataAtRestKeyProvider} in
 * src/AetherNet.Storage/DerivedDataAtRestKeyProvider.cs (which uses PBKDF2
 * 600k — chosen because the .NET BCL ships PBKDF2 but not scrypt).
 * TypeScript uses scrypt because Node's BCL ships scrypt but only ships
 * PBKDF2 via the much-slower software path; both reach the OWASP
 * recommendation for their respective KDF.
 *
 * SPDX-License-Identifier: MIT
 */
import { scrypt } from "node:crypto";
import { DataAtRestKeyProvider } from "./DataAtRestKeyProvider.js";

/**
 * OWASP 2023 recommendation for scrypt: N=2^14, r=8, p=1. The cost
 * parameter is just N here; r and p are pinned to the OWASP-recommended
 * defaults inside the provider.
 */
export const DEFAULT_DERIVED_KEY_COST = 16384;

const KEY_BYTE_LENGTH = 32; // AES-256
const MINIMUM_SALT_LENGTH = 16;
const SCRYPT_R = 8;
const SCRYPT_P = 1;

/**
 * Per-version derivation parameters captured at construction time.
 */
interface VersionParams {
  passphrase: string;
  salt: Uint8Array;
  cost: number;
}

export class DerivedDataAtRestKeyProvider implements DataAtRestKeyProvider {
  /**
   * Resolved keys, keyed by version. Filled lazily on first
   * {@link getKey} for that version (or eagerly via {@link prime}).
   */
  private readonly cachedKeys: Map<number, Uint8Array> = new Map();
  private readonly versionParams: Map<number, VersionParams>;
  readonly currentVersion: number;
  /** The default cost the provider was constructed with. */
  readonly cost: number;

  /**
   * Construct a single-version provider that derives version 1 from the
   * supplied passphrase and salt. The first {@link getKey} call (or
   * {@link prime}) actually performs the scrypt computation; subsequent
   * calls hit the cache.
   */
  static async create(
    passphrase: string,
    salt: Uint8Array,
    cost: number = DEFAULT_DERIVED_KEY_COST
  ): Promise<DerivedDataAtRestKeyProvider> {
    validateInputs(passphrase, salt, cost);
    const params = new Map<number, VersionParams>();
    params.set(1, { passphrase, salt: new Uint8Array(salt), cost });
    const provider = new DerivedDataAtRestKeyProvider(params, 1, cost);
    await provider.prime(1);
    return provider;
  }

  /**
   * Synchronous variant that defers the scrypt cost to the first
   * {@link getKey} call. Useful if you can't await at construction time
   * — but be aware {@link getKey} returns null until {@link prime} has
   * been awaited at least once for the relevant version.
   */
  static lazy(
    passphrase: string,
    salt: Uint8Array,
    cost: number = DEFAULT_DERIVED_KEY_COST
  ): DerivedDataAtRestKeyProvider {
    validateInputs(passphrase, salt, cost);
    const params = new Map<number, VersionParams>();
    params.set(1, { passphrase, salt: new Uint8Array(salt), cost });
    return new DerivedDataAtRestKeyProvider(params, 1, cost);
  }

  private constructor(
    versionParams: Map<number, VersionParams>,
    currentVersion: number,
    cost: number
  ) {
    this.versionParams = versionParams;
    this.currentVersion = currentVersion;
    this.cost = cost;
  }

  /**
   * Forces derivation of the key for {@code version}, populating the
   * internal cache. Idempotent.
   */
  async prime(version: number): Promise<void> {
    if (this.cachedKeys.has(version)) return;
    const params = this.versionParams.get(version);
    if (!params) {
      throw new Error(`No derivation parameters registered for version ${version}.`);
    }
    const derived = await derive(params.passphrase, params.salt, params.cost);
    this.cachedKeys.set(version, derived);
  }

  getKey(version: number): Uint8Array | null {
    return this.cachedKeys.get(version) ?? null;
  }

  /**
   * Returns a new provider that adds a freshly-derived key under
   * {@code newVersion} (which becomes the new current version) while
   * keeping every existing version available for decryption. The new
   * key is primed before the promise resolves.
   */
  async withRotation(
    newVersion: number,
    newPassphrase: string,
    newSalt: Uint8Array,
    cost?: number
  ): Promise<DerivedDataAtRestKeyProvider> {
    if (!Number.isInteger(newVersion) || newVersion < 1 || newVersion > 255) {
      throw new Error(`newVersion must be in [1, 255] (got ${newVersion}).`);
    }
    if (this.versionParams.has(newVersion)) {
      throw new Error(`Version ${newVersion} already exists in this provider.`);
    }
    const effectiveCost = cost ?? this.cost;
    validateInputs(newPassphrase, newSalt, effectiveCost);

    const params = new Map<number, VersionParams>(this.versionParams);
    params.set(newVersion, {
      passphrase: newPassphrase,
      salt: new Uint8Array(newSalt),
      cost: effectiveCost,
    });

    const next = new DerivedDataAtRestKeyProvider(params, newVersion, effectiveCost);
    // Carry over already-derived keys so we don't redo the scrypt work.
    for (const [v, k] of this.cachedKeys.entries()) {
      next.cachedKeys.set(v, k);
    }
    await next.prime(newVersion);
    return next;
  }
}

function validateInputs(passphrase: string, salt: Uint8Array, cost: number): void {
  if (!passphrase) throw new Error("passphrase cannot be empty");
  if (!salt) throw new Error("salt cannot be null");
  if (salt.length < MINIMUM_SALT_LENGTH) {
    throw new Error(`Salt must be at least ${MINIMUM_SALT_LENGTH} bytes.`);
  }
  if (!Number.isInteger(cost) || cost < 2) {
    throw new Error("cost must be an integer >= 2.");
  }
  // scrypt N must be a power of 2.
  if ((cost & (cost - 1)) !== 0) {
    throw new Error(`cost must be a power of 2 (got ${cost}).`);
  }
}

function derive(
  passphrase: string,
  salt: Uint8Array,
  cost: number
): Promise<Uint8Array> {
  // Node's scrypt has a default maxmem cap that 2^14 fits comfortably
  // under (~16 MiB at r=8). For higher costs callers can raise this —
  // for the OWASP-recommended cost of 2^14 the default is fine.
  return new Promise((resolve, reject) => {
    scrypt(
      passphrase,
      salt,
      KEY_BYTE_LENGTH,
      { N: cost, r: SCRYPT_R, p: SCRYPT_P },
      (err, derivedKey) => {
        if (err) reject(err);
        else resolve(new Uint8Array(derivedKey));
      }
    );
  });
}
